### PR TITLE
ping: Set mark (SO_MARK) on probe socket

### PR DESCRIPTION
**Pull request by PaulSD**
If the bind (source) IP is is not explicitly specified with "-I" then
ping creates a "probe" socket to determine the source IP.  However, if a
mark is configured with "-m", the current code does not apply this mark
to the "probe" socket.  If policy routing is configured such that the
mark affects source IP selection, then the wrong source IP will be used.
In addition, if policy routing is configured such that the destination
will be unreachable unless the mark is set, then the probe socket will
fail and ping will exit with an error.

This commit sets the mark on the probe socket in addition to the actual
ping socket.

### CODE REVIEW

**Comment by pevik** [comment]
@PaulSD I'm sorry for a delay with reply. Code LGTM, it looks as a valid issue. But are you sure that applying mark to probe socket does not break anything? Could you please share steps to reproduce the issue? I looked into `tcpdump` output, but don't see the mark there.

Also, shouldn't `sock_setmark(rts, probe_fd);` be added also into `ping6_run()`? 

https://github.com/iputils/iputils/blob/f0e097e6bedb1b90a783560ff766a0045a394da0/ping/ping6_common.c#L181-L187

**Comment by pevik** [comment]
> @PaulSD I'm sorry for a delay with reply. Code LGTM, it looks as a valid issue. But are you sure that applying mark to probe socket does not break anything? 

I plan to merge slightly modified code, see my branch https://github.com/pevik/iputils/commits/gh%23340.SO_MARK.fixes
(currently https://github.com/pevik/iputils/commit/1c4678fd630c75227a88618779290636469c1a4e). Changes:
* move if (rts->opt_mark) out of the function
* pass `unsigned int mark` instead of `struct ping_rts` (it might be `int` if we don't merge https://github.com/iputils/iputils/pull/345)
* add warning `error(0, errno_save, _("WARNING: SO_MARK not supported"));`

> Could you please share steps to reproduce the issue? 
It'd be nice if you share the setup, so that I could later create test for it.

> I looked into `tcpdump` output, but don't see the mark there.

I see `SO_MARK` is "not on the wire" thus cannot be seen in with `tcpdump`.

> Also, shouldn't `sock_setmark(rts, probe_fd);` be added also into `ping6_run()`?

I see was added also to `ping6_run()`. I verified it's working for both IPv4 and IPv6.

Whole diff to what I'm commit to merge:
```
diff --git ping/ping.c ping/ping.c
index a709c33..b24a2cc 100644
--- ping/ping.c
+++ ping/ping.c
@@ -714,7 +714,9 @@ int ping4_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 		if (rts->settos &&
 		    setsockopt(probe_fd, IPPROTO_IP, IP_TOS, (char *)&rts->settos, sizeof(int)) < 0)
 			error(0, errno, _("warning: QOS sockopts"));
-		sock_setmark(rts, probe_fd);
+
+		if (rts->opt_mark)
+			sock_setmark(rts->mark, probe_fd);
 
 		dst.sin_port = htons(1025);
 		if (rts->nroute)
diff --git ping/ping.h ping/ping.h
index 21d1699..e058ac4 100644
--- ping/ping.h
+++ ping/ping.h
@@ -380,7 +380,7 @@ char *pr_addr(struct ping_rts *rts, void *sa, socklen_t salen);
 int is_ours(struct ping_rts *rts, socket_st *sock, uint16_t id);
 extern int pinger(struct ping_rts *rts, ping_func_set_st *fset, socket_st *sock);
 extern void sock_setbufs(struct ping_rts *rts, socket_st *, int alloc);
-extern void sock_setmark(struct ping_rts *rts, int sock_fd);
+extern void sock_setmark(unsigned int mark, int fd);
 extern void setup(struct ping_rts *rts, socket_st *);
 extern int main_loop(struct ping_rts *rts, ping_func_set_st *fset, socket_st*,
 		     uint8_t *packet, int packlen);
diff --git ping/ping6_common.c ping/ping6_common.c
index e9499b7..13bf45a 100644
--- ping/ping6_common.c
+++ ping/ping6_common.c
@@ -183,7 +183,8 @@ int ping6_run(struct ping_rts *rts, int argc, char **argv, struct addrinfo *ai,
 		    !IN6_IS_ADDR_MC_LINKLOCAL(&rts->firsthop.sin6_addr))
 			rts->firsthop.sin6_family = AF_INET6;
 
-		sock_setmark(rts, probe_fd);
+		if (rts->opt_mark)
+			sock_setmark(rts->mark, probe_fd);
 
 		rts->firsthop.sin6_port = htons(1025);
 		if (connect(probe_fd, (struct sockaddr *)&rts->firsthop, sizeof(rts->firsthop)) == -1) {
diff --git ping/ping_common.c ping/ping_common.c
index 7b9850e..1a61918 100644
--- ping/ping_common.c
+++ ping/ping_common.c
@@ -446,25 +446,22 @@ void sock_setbufs(struct ping_rts *rts, socket_st *sock, int alloc)
 	}
 }
 
-/* Set SO_MARK socket option. */
-
-void sock_setmark(struct ping_rts *rts, int sock_fd)
+void sock_setmark(unsigned int mark, int fd)
 {
 #ifdef SO_MARK
-	if (rts->opt_mark) {
-		int ret;
-		int errno_save;
-
-		enable_capability_admin();
-		ret = setsockopt(sock_fd, SOL_SOCKET, SO_MARK, &rts->mark, sizeof(rts->mark));
-		errno_save = errno;
-		disable_capability_admin();
-
-		if (ret == -1) {
-			/* Do not exit, old kernels do not support mark. */
-			error(0, errno_save, _("Warning: Failed to set mark: %d"), rts->mark);
-		}
-	}
+	int ret;
+	int errno_save;
+
+	enable_capability_admin();
+	ret = setsockopt(fd, SOL_SOCKET, SO_MARK, mark, sizeof(mark));
+	errno_save = errno;
+	disable_capability_admin();
+
+	/* Do not exit, old kernels do not support mark. */
+	if (ret == -1)
+		error(0, errno_save, _("WARNING: failed to set mark: %d"), mark);
+#else
+		error(0, errno_save, _("WARNING: SO_MARK not supported"));
 #endif
 }
 
@@ -498,7 +495,9 @@ void setup(struct ping_rts *rts, socket_st *sock)
 			error(0, 0, _("Warning: no SO_TIMESTAMP support, falling back to SIOCGSTAMP"));
 	}
 #endif
-	sock_setmark(rts, sock->fd);
+
+	if (rts->opt_mark)
+		sock_setmark(rts->mark, sock->fd);
 
 	/* Set some SNDTIMEO to prevent blocking forever
 	 * on sends, when device is too slow or stalls. Just put limit
```

**Comment by PaulSD** [comment]
Sorry, I'm not ignoring this - I've had it near the top of my to-do list for months, and just never got a moment to assemble a thoughtful response.  I'll try to take a closer look this evening.

> But are you sure that applying mark to probe socket does not break anything?
Yes, I'm confident it doesn't break anything.

> Could you please share steps to reproduce the issue?
Yes, I will put together a simple test case.

**Comment by PaulSD** [comment]
Your modifications LGTM.

Simple test case:
```
$ ip link set dev eth0 up
$ ip addr add 192.168.0.2/24 dev eth0
$ ip rule add priority 101 fwmark 0x2/0x1 table 1
$ ip route add table 1 default via 192.168.0.1 dev eth0
$ old-ping -c 1 -m 1 1.1.1.1
ping: connect: Network is unreachable
$ new-ping -c 1 -m 1 1.1.1.1
<success>
$ ip route add default via 192.168.0.1 dev eth0
$ old-ping -c 1 -m 1 1.1.1.1
<success>
```

For reference:
I ran into this issue on a system with two links to the Internet via two different ISPs.  I set up policy routing similar to the above to allow me to use SO_MARK to explicitly select the ISP that should be used for each outbound packet/connection.  Then I wrote a monitoring script that would use `ping` (among other things) to monitor each link and decide which link should be the primary/default.  If no default route was configured (if a mark was required for routing to work, for example immediately after the system boots up and before the monitoring script has decided which ISP to use) then `ping` would fail.

**Comment by pevik** [comment]
Thanks! I'll try to transform it to `netns` to be usable for testing.

**Comment by pevik** [comment]
Hm, I wonder how to setup routing with 2 different paths. Because simple testcase with defining network namespaces and delete route + adding route via mark works on unpatched ping:
```
IPUTILS_NS="ip netns exec iputils_ns"
ip net add iputils_ns
ip link add name iputils_veth1 type veth peer name iputils_veth2
ip link set dev iputils_veth1 netns iputils_ns
$IPUTILS_NS ip link set lo up

ip xfrm policy flush
ip xfrm state flush
ip link set iputils_veth2 down
ip route flush dev iputils_veth2
ip addr flush dev iputils_veth2
ip link set iputils_veth2 up
ip addr add 10.0.0.2/24 dev iputils_veth2

$IPUTILS_NS ip xfrm policy flush
$IPUTILS_NS ip xfrm state flush
$IPUTILS_NS ip link set iputils_veth1 down
$IPUTILS_NS ip route flush dev iputils_veth1
$IPUTILS_NS ip addr flush dev iputils_veth1
$IPUTILS_NS ip link set iputils_veth1 up
$IPUTILS_NS ip addr add 10.0.0.1/24 dev iputils_veth1

ip route
10.0.0.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2

# delete default route
ip route del 10.0.0.0/24 dev iputils_veth2

ping 10.0.0.1 -c1
From 10.156.163.92 icmp_seq=1 Destination Host Unreachable
ping 10.0.0.1 -c1 -m1
From 10.156.163.92 icmp_seq=1 Destination Host Unreachable

ip rule add priority 101 fwmark 0x1/0x1 table 1
ip route add table 1 10.0.0.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2

ping 10.0.0.1 -c1
From 10.156.163.92 icmp_seq=1 Destination Host Unreachable
ping 10.0.0.1 -c1 -m1
64 bytes from 10.0.0.1: icmp_seq=1 ttl=64 time=0.048 ms
```

**Comment by PaulSD** [comment]
The problem is that you are pinging an IP on the local subnet, which changes the behavior in a number of subtle ways.

Try this:
```
ip addr add 10.0.0.2/24 dev iputils_veth2
$IPUTILS_NS ip addr add 10.0.0.1/24 dev iputils_veth1
$IPUTILS_NS ip addr add 10.0.1.1/24 dev iputils_veth1

ip rule add priority 101 fwmark 0x1/0x1 table 1
ip route add table 1 10.0.1.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2

ping 10.0.1.1 -c1  # This should fail
ping 10.0.1.1 -c1 -m1  # This should succeed but actually fails
```

**Comment by pevik** [comment]
Thanks, I'll test it later this week.

**Comment by pevik** [comment]
@PaulSD Hm, it still works also with the old version. I need to have deeper look into route setup. But feel free to find working solution (if you find it, please post whole setup).

**Comment by PaulSD** [comment]
*scratches head*
```
# IPUTILS_NS="ip netns exec iputils_ns"
# ip net add iputils_ns
# ip link add name iputils_veth1 type veth peer name iputils_veth2 
# ip link set dev iputils_veth1 netns iputils_ns
# ip link set iputils_veth2 up
# ip addr add 10.0.0.2/24 dev iputils_veth2
# $IPUTILS_NS ip link set iputils_veth1 up
# $IPUTILS_NS ip addr add 10.0.0.1/24 dev iputils_veth1
# $IPUTILS_NS ip addr add 10.0.1.1/24 dev iputils_veth1
# ip rule add priority 101 fwmark 0x1/0x1 table 1
# ip route add table 1 10.0.1.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2 
# ping 10.0.1.1 -c1
PING 10.0.1.1 (10.0.1.1) 56(84) bytes of data.

--- 10.0.1.1 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms

# ping 10.0.1.1 -c1 -m1
PING 10.0.1.1 (10.0.1.1) 56(84) bytes of data.

--- 10.0.1.1 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms

# ip route add 10.0.1.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2 
# ping 10.0.1.1 -c1
PING 10.0.1.1 (10.0.1.1) 56(84) bytes of data.
64 bytes from 10.0.1.1: icmp_seq=1 ttl=64 time=0.163 ms

--- 10.0.1.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.163/0.163/0.163/0.000 ms
# ping -V
ping from iputils 20210202-85-gf9fb573

**Comment by pevik** [comment]
Again, in my case `./builddir/ping/ping 10.0.1.1 -c1 -m1` works even without your patch (I use `20210722-41-gf9fb573`, I suppose it's the same as yours `20210202-85-gf9fb573` - master from previous git release?).

I suspect there is something different on your system. Could you run `sysctl -a | grep net` ?

I also have everything `fwmark` related disabled, also `ping_group_range` is interesting (that should be somehow related to `ICMP_PROTO` sockets for all users instead of `SOCK_RAW` sockets, which use `cap_net_raw+p`):
```
# sysctl -a |grep -e fwmark -e ping_group_range
net.ipv4.fwmark_reflect = 0
net.ipv4.ping_group_range = 1	0
net.ipv4.tcp_fwmark_accept = 0
net.ipv6.fwmark_reflect = 0

```

**Comment by PaulSD** [comment]
That is very weird, and doesn't make any sense to me.  Without my changes, the existing logic should definitely result in a failure in this case.

I'm using a fairly stock Ubuntu 20.04, and I originally discovered this on a fairly stock Ubuntu 18.04 system (and my patch fixed it there too).
[sysctls.txt](https://github.com/iputils/iputils/files/7576597/sysctls.txt) are attached.  I do have net.ipv4.conf.all.forwarding=1, but I don't think that should affect this.  I don't believe I've changed anything else that might be relevant.

Re: `20210202-85-gf9fb573` - I cloned master a few months ago, built it, then a few days ago did a `git pull` and re-built it again with `./build.sh`.  Seems like it updated the commit ID part of the version number but not the date?  Since the Git commits match, I think we are actually running the same thing, assuming it isn't caching something somewhere...

**Comment by pevik** [comment]
Indeed, `net.ipv4.conf.all.forwarding=1` is not related. Also `net.ipv4.ping_group_range="0   2147483647"` is not related either. You also don't have `iputils_veth[12]` devices in the output, I suppose you use `ens3` for testing, right?

Do you test with root or with regular user? There is permission issue on Debian stable (which I don't think is your issue):
```
# sysctl net.ipv4.ping_group_range="1 1000" # ping for user
$ ./builddir/ping/ping ::1 -c1
PING ::1(::1) 56 data bytes
64 bytes from ::1: icmp_seq=1 ttl=64 time=0.070 ms

--- ::1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.068/0.068/0.068/0.000 ms

$ ./builddir/ping/ping 10.0.1.1 -c1 -m1
PING 10.0.1.1 (10.0.1.1) 56(84) bytes of data.
./builddir/ping/ping: Warning: Failed to set mark: 1: Operation not permitted

--- 10.0.1.1 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```

> Re: 20210202-85-gf9fb573 - I cloned master a few months ago, built it, then a few days ago did a git pull and re-built it again with ./build.sh. Seems like it updated the commit ID part of the version number but not the date? Since the Git commits match, I think we are actually running the same thing, assuming it isn't caching something somewhere...

Git does not cache anything. Your code is based on 4f3ec1da98d4159dd24278d8598da6d25b1543d2, thus you either didn't push changes with `git push -f  PaulSD` or you haven't rebased properly, e.g. `git pull -r`.

I rebased your code in my PR: https://github.com/pevik/iputils/commits/gh%23340.SO_MARK.fixes. Although there quite a lot ping related changes (e.g. 3aca3574a93f52fcbf6def5ddb45cff9d4a8a2e5, f52b582248f1f870e870a9973621805d969906b4, 7c65999f98bc4a1984594b7fad1af0eaf0b9d34b, 5f6bec5ab57cc8beaa78f5756a0ffbdf01f28d36, 79d713eab6181e219bf932b404706f6f59ff2539), the result is the same on this old release - one more set mark operation (not sure when it was removed):
```
./builddir/ping/ping 10.0.1.1 -c1 -m1
./builddir/ping/ping: Warning: Failed to set mark: 1: Operation not permitted
PING 10.0.1.1 (10.0.1.1) 56(84) bytes of data.
./builddir/ping/ping: Warning: Failed to set mark: 1: Operation not permitted

--- 10.0.1.1 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```

**Comment by PaulSD** [comment]
> I suppose you use ens3 for testing, right?
For testing, I created iputils_veth, then deleted it immediately after testing.  I can dump sysctls after testing and before deleting if that would help...

> Do you test with root or with regular user?
With root.

>  thus you either didn't push changes with `git push -f PaulSD`
Correct, I haven't pushed changes...  Basically, several months ago I did `git clone` and got commit `4f3ec1d`.  I built that and tested and confirmed that it behaved the same as the `ping` that came with Ubuntu.  Then I wrote my patch and tested, then pushed to GitHub.  Most of my testing was with those two binaries.
Then on 11/18/2021, since you said you couldn't reproduce, I tried: `git fetch ; git log origin/master ; git checkout f9fb57389b2cbefba9f0532a2c6c4453279912da ; ./build.sh`

Let me try a fresh VM.

**Comment by pevik** [comment]
> > I suppose you use ens3 for testing, right?
> 
> For testing, I created iputils_veth, then deleted it immediately after testing. I can dump sysctls after testing and before deleting if that would help...

Yes please. Could you please reproduce the problem on fresh distro (independent on your particular setup on currently affected machine) and post the steps? I'd like to be able to reproduce the issue before merging it. Although this looks safe, there are few corner cases in ping thus I hesitate to merge things which "should work". Also this is the last thing which block the release.

**Comment by PaulSD** [comment]
Doh!  I see now.
The test case was missing an important piece, so it wasn't an effective test case.  This was only failing during my testing due to what appears to be a bug in the iptables `rpfilter` extension.

To properly test this, we need to add `ip route add unreachable 10.0.1.0/24`:
```
IPUTILS_NS="ip netns exec iputils_ns"
ip net add iputils_ns
ip link add name iputils_veth1 type veth peer name iputils_veth2 
ip link set dev iputils_veth1 netns iputils_ns
ip link set iputils_veth2 up
ip addr add 10.0.0.2/24 dev iputils_veth2
$IPUTILS_NS ip link set iputils_veth1 up
$IPUTILS_NS ip addr add 10.0.0.1/24 dev iputils_veth1
$IPUTILS_NS ip addr add 10.0.1.1/24 dev iputils_veth1
ip route add unreachable 10.0.1.0/24
ip rule add priority 101 fwmark 0x1/0x1 table 1
ip route add table 1 10.0.1.0/24 dev iputils_veth2 proto kernel scope link src 10.0.0.2 

old_ping 10.0.1.1 -c1 -m1  # This should fail with "ping: connect: No route to host"
new_ping 10.0.1.1 -c1 -m1  # This should succeed
```

Details:
* Prior to this PR, the lack of the mark on the probe socket is only really a problem if it results in the probe socket getting back a "host unreachable" or similar error (eg. `EINVAL` due to hitting a blackhole route, or there are several other similar possibilities).  If the probe socket does not get an error on `connect()` then it will set `rts->source` to the wrong source address (set to the source address that would be used without the mark, when it should be set to the source address that would be used with the mark).  However, since `bind()` is only called with an explicit source address if `opt_strictsource` is true, the wrong value in `rts->source` doesn't get used, and everything works properly despite the incorrect `rts->source` value.
** Note that this means my commit message (and my first comment in this PR) is partially incorrect - the missing mark on the probe socket will not result in the wrong source address being used.  It will only result in `rts->source` being incorrectly populated, although that doesn't really matter since that value doesn't get used in this case.  (I'll fix my commit message.)
* Therefore, prior to this PR, if the destination matches any valid non-mark route (like your normal default route), then the probe socket will not return an error and ping will work fine.
* As I mentioned earlier, the original use case where I ran into problems involved a system that might not have a default route at all times, and therefore might require the mark to be set on the probe socket in order to match any route and avoid an "unreachable" error.  Note that my first test case in this PR involved deleting the default route, which properly tested this.
* Of course, deleting the default route might be disruptive to the test system.  So, when we started playing with namespaces, I figured I would come up with something that would test the `rts->source` mis-configuration instead of deleting the default route to test the "unreachable" error behavior.  However, as explained here, I now know that mis-configuration doesn't actually affect the behavior of `ping`, so my namespace-based test case wasn't really valid.
* To make this test case valid, we can simply add an explicit "unreachable" route for the ping destination, which properly tests the "unreachable" error behavior without deleting the default route.
* The reason `ping` was failing for me when I attempted to test the `rts->source` mis-configuration is because `iptables --table mangle --append PREROUTING --match rpfilter --validmark --jump ...` seems to have a bug where it doesn't properly match marked reply traffic for `ping`, which in my case results in the reply packets being dropped.  This mislead me into thinking that `ping` was setting the wrong source address and traffic was being lost because of that.  In reality, `ping` was working as expected (using the correct source address) and it was actually `rpfilter` that was misbehaving.

**Comment by pevik** [comment]
FYI yes, this works, but only for `SOCK_RAW` via `CAP_NET_RAW`, which is e.g on Debian (and probably Ubuntu):

https://github.com/iputils/iputils/blob/f9fb57389b2cbefba9f0532a2c6c4453279912da/ping/ping.c#L799-L808

Recent openSUSE and Fedora, which use `ICMP_PROTO` (https://bugzilla.suse.com/show_bug.cgi?id=1174504, requires `sysctl net.ipv4.ping_group_range = 0	2147483647`) and it does not work on that:

```
# ./builddir/ping/ping 10.0.1.1 -c1 -m1
sendto(3, "\10\0\241\201\0\0\0\1O\37\252a\0\0\0\0\223)\v\0\0\0\0\0\20\21\22\23\24\25\26\27"..., 64, 0, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("10.0.1.1")}, 16) = 64
setitimer(ITIMER_REAL, {it_interval={tv_sec=0, tv_usec=0}, it_value={tv_sec=10, tv_usec=0}}, NULL) = 0
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EAGAIN (Resource temporarily unavailable)
recvmsg(3, {msg_namelen=128}, 0)        = -1 EINTR (Interrupted system call)
--- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[]})                 = -1 EINTR (Interrupted system call)
write(1, "\n", 1)                       = 1
write(1, "--- 10.0.1.1 ping statistics ---"..., 33) = 33
write(1, "1 packets transmitted, 0 receive"..., 62) = 62
write(1, "\n", 1)                       = 1
close(1)                                = 0
close(2)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by PaulSD** [comment]
You're saying ping is setting the SO_MARK socket option without getting any errors but then the kernel isn't using it?  Or are you getting permission errors setting SO_MARK?

I think SO_MARK requires some caps, so permission errors wouldn't be surprising if caps have been dropped from the ping command...

**Comment by pevik** [comment]
Receiving packets is broken on last setup with `ip route add unreachable 10.0.1.0/24` if not using socket with `SOCK_RAW`. I *think* it's enabled on `SOCK_RAW` via `ICMP_FILTER`, but not sure (just comparing strace output of both cases). It'd be interesting for me to know whether this fix is working on your real setup even without `SOCK_RAW` (i.e. using `SOCK_DGRAM` and not falling back to `SOCK_RAW` in `create_socket()`:

https://github.com/iputils/iputils/blob/f9fb57389b2cbefba9f0532a2c6c4453279912da/ping/ping.c#L129-L147

Anyway I'm going to accept this PR without further changes, because we probably made testing a bit arbitrary with `ip route add unreachable 10.0.1.0/24` and the main question is not whether your corner case is always working (although preferable), but whether it didn't introduce any regression (didn't break anything). But you can see how socket networking is quite complicated and nobody from current maintainers know deeply all the socket options consequences (there has been a lot regressions caused by fixes).

**Comment by PaulSD** [comment]
Both `SOCK_RAW` and `ICMP_PROTO` seem to work properly for me on Ubuntu 20.04 when running as root with this patch using the test case above (also works with my real-world use case):
```
# echo '1 0' > /proc/sys/net/ipv4/ping_group_range
# strace -o trace.log ./builddir/ping/ping 10.0.1.1 -c1 -m1
socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP) = -1 EACCES (Permission denied)
socket(AF_INET, SOCK_RAW, IPPROTO_ICMP) = 3
setsockopt(3, SOL_SOCKET, SO_MARK, [1], 4) = 0
sendto(3, ...
recvmsg(3, {msg_name=...
64 bytes from 10.0.1.1: icmp_seq=1 ttl=64 time=0.092 ms
# echo '0 2147483647' > /proc/sys/net/ipv4/ping_group_range
# strace -o trace.log ./builddir/ping/ping 10.0.1.1 -c1 -m1
socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP) = 3
setsockopt(3, SOL_SOCKET, SO_MARK, [1], 4) = 0
sendto(3, ...
recvmsg(3, {msg_name=...
64 bytes from 10.0.1.1: icmp_seq=1 ttl=64 time=0.086 ms
```

`SO_MARK` requires `CAP_NET_ADMIN`, so when running as non-root without caps it is expected to fail:
```
$ ./ping 10.0.1.1 -c1 -m1
./ping: WARNING: failed to set mark: 1: Operation not permitted
./ping: connect: No route to host
```

**Comment by pevik** [comment]
> `SO_MARK` requires `CAP_NET_ADMIN`, so when running as non-root without caps it is expected to fail:
> 
> ```
> $ ./ping 10.0.1.1 -c1 -m1
> ./ping: WARNING: failed to set mark: 1: Operation not permitted
> ./ping: connect: No route to host
> ```

```
FYI: ICMP ping requires special setup:
# sysctl -w net.ipv4.ping_group_range='0 2147483647'
$ meson configure builddir -DNO_SETCAP_OR_SUID=false && make
$ ./builddir/ping/ping github.com -c1
PING github.com (140.82.121.4) 56(84) bytes of data.
64 bytes from lb-140-82-121-4-fra.github.com (140.82.121.4): icmp_seq=1 ttl=49 time=20.3 ms

--- github.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 20.267/20.267/20.267/0.000 ms
```

**Comment by pevik** [comment]
Testing on various distros Debian stable, Debian stable (bullseye), openSUSE Tumbleweed and Fedora 34, Ubuntu Focal (current LTS), Ubuntu Hirsute:
* forcing ICMP socket `SOCK_DGRAM` (with `sysctl -w net.ipv4.ping_group_range='0 2147483647'`):
it works only on Debian stable (bullseye) and current testing (bookworm), fails on other distros

* forcing `SOCK_RAW` (with `sysctl -w net.ipv4.ping_group_range='1 0'`):
	again working only on Debian stable (bullseye) and current testing (bookworm), fails on other distros

IMHO if you test on fresh VM or docker you would see the same result.

=> different distro network related configuration/setup?
I don't see any difference in `sysctl -a |grep net`. And it does not look to be kernel / glibc version related.

Just to verify packets are being sent on all tested distros:
```
tcpdump icmp -vv -i any -y LINUX_SLL2
10:25:39.352689 iputils_veth2 Out IP (tos 0x0, ttl 64, id 8892, offset 0, flags [DF], proto ICMP (1), length 84)
    10.0.0.2 > 10.0.1.1: ICMP echo request, id 62606, seq 1, length 64
10:25:39.352748 iputils_veth2 In  IP (tos 0x0, ttl 64, id 1574, offset 0, flags [none], proto ICMP (1), length 84)
    10.0.1.1 > 10.0.0.2: ICMP echo reply, id 62606, seq 1, length 64
```

**Comment by PaulSD** [comment]
> FYI: ICMP ping requires special setup:
> 
> ```
> # sysctl -w net.ipv4.ping_group_range='0 2147483647'
> $ meson configure builddir -DNO_SETCAP_OR_SUID=false && make
> $ ./builddir/ping/ping github.com -c1
> PING github.com (140.82.121.4) 56(84) bytes of data.
> 64 bytes from lb-140-82-121-4-fra.github.com (140.82.121.4): icmp_seq=1 ttl=49 time=20.3 ms
> 
> --- github.com ping statistics ---
> 1 packets transmitted, 1 received, 0% packet loss, time 0ms
> rtt min/avg/max/mdev = 20.267/20.267/20.267/0.000 ms
> ```

Yeah, I did that.  ICMP ping worked fine when running without caps/root ... but SO_MARK does not work without caps/root (regardless of whether SOCK_RAW or IPPROTO_ICMP is used).

I'll try again on some fresh VMs and see if there is something else we are missing here...

**Comment by pevik** [comment]
@PaulSD merged. Thanks a lot for your effort!
